x/rollapp/keeper/fraud_handler.go

```diff
@@ -1,56 +1,44 @@
 package keeper
 
 import (
-	"errors"
 	"fmt"
 
 	errorsmod "cosmossdk.io/errors"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
-	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
-	cometbfttypes "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint"
-
 	common "github.com/dymensionxyz/dymension/v3/x/common/types"
 	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
 )
 
-// HandleFraud handles the fraud evidence submitted by the user.
-func (k Keeper) HandleFraud(ctx sdk.Context, rollappID, clientID string, fraudHeight uint64, seqAddr string) error {
-	stateInfo, err := k.FindStateInfoByHeight(ctx, rollappID, fraudHeight)
-	if err != nil {
-		return err
-	}
-
-	// check height is not finalized
-	if stateInfo.Status == common.Status_FINALIZED {
-		return errorsmod.Wrapf(types.ErrDisputeAlreadyFinalized, "state info for height %d is already finalized", fraudHeight)
+// HardFork handles the fraud evidence submitted by the user.
+func (k Keeper) HardFork(ctx sdk.Context, rollappID string, fraudHeight uint64) error {
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return gerrc.ErrNotFound
 	}
 
-	// check height is not reverted
-	if stateInfo.Status == common.Status_REVERTED {
-		return errorsmod.Wrapf(types.ErrDisputeAlreadyReverted, "state info for height %d is already reverted", fraudHeight)
-	}
-
-	// check the sequencer for this height is the same as the one in the fraud evidence
-	if stateInfo.Sequencer != seqAddr {
-		return errorsmod.Wrapf(types.ErrWrongProposerAddr, "sequencer address %s does not match the one in the state info", seqAddr)
-	}
+	// FIXME: enable later on
+	/*
+		// enforce the assumption that genesis transfer is completed before hard fork
+		if !rollapp.IsTransferEnabled() {
+			return errorsmod.Wrapf(gerrc.ErrFailedPrecondition, "can't hard fork before transfer is enabled")
+		}
+	*/
 
-	// check that the clientID is correct
-	err = k.verifyClientID(ctx, rollappID, clientID)
+	err := k.RevertPendingStates(ctx, rollappID, fraudHeight)
 	if err != nil {
-		return errors.Join(types.ErrWrongClientId, err)
+		return errorsmod.Wrap(err, "revert pending states")
 	}
 
-	// freeze the rollapp and revert all pending states
-	err = k.FreezeRollapp(ctx, rollappID)
-	if err != nil {
-		return fmt.Errorf("freeze rollapp: %w", err)
-	}
+	// update revision number
+	rollapp.RevisionNumber += 1
+	// FIXME: set liveness height
```

Comment:
yeah good point
I wouldn't worry about it too much until we integrate both our prs